Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracksExpress_cff.py

```diff
@@ -10,19 +10,30 @@
 from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *
 from RecoPPS.Configuration.recoCTPPS_cff import *
 
-totemTriggerRawToDigi.rawDataTag = cms.InputTag('hltPPSCalibrationRaw')
-totemRPRawToDigi.rawDataTag = cms.InputTag('hltPPSCalibrationRaw')
-ctppsDiamondRawToDigi.rawDataTag = cms.InputTag('hltPPSCalibrationRaw')
-totemTimingRawToDigi.rawDataTag = cms.InputTag('hltPPSCalibrationRaw')
-ctppsPixelDigis.inputLabel = cms.InputTag('hltPPSCalibrationRaw')
-
-ctppsRawToDigiTask = cms.Task(
-  totemTriggerRawToDigi,
-  totemRPRawToDigi,
-  ctppsDiamondRawToDigi,
-  totemTimingRawToDigi,
-  ctppsPixelDigis
+
+_totemTriggerRawToDigi = totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
+_totemTriggerRawToDigi.setLabel("totemTriggerRawToDigi")
```

Comment:
The recommended pattern would be along
```suggestion
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTriggerRawToDigi as _totemTriggerRawToDigi
totemTriggerRawToDigi = _totemTriggerRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
```

On the top of my head I'm not even sure what would happen here.

Reply:
If I follow your suggestion, namely doing:

```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTriggerRawToDigi as _totemTriggerRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemRPRawToDigi as _totemRPRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsDiamondRawToDigi as _ctppsDiamondRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis

from RecoPPS.Configuration.recoCTPPS_cff import *

totemTriggerRawToDigi = _totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemRPRawToDigi = _totemRPRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsDiamondRawToDigi = _ctppsDiamondRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemTimingRawToDigi = _totemTimingRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsPixelDigis = _ctppsPixelDigis.clone(inputLabel = cms.InputTag('hltPPSCalibrationRaw'))

_ctppsRawToDigiTask = cms.Task(
  totemTriggerRawToDigi,
  totemRPRawToDigi,
  ctppsDiamondRawToDigi,
  totemTimingRawToDigi,
  ctppsPixelDigis
)
```

then I get the following error:


```python
ValueError: Trying to override definition of ctppsDiamondRawToDigi while it is used by the task RawToDigiTask
 new object defined in: /tmp/lgrzanka/CMSSW_12_3_X_2022-01-23-2300/python/Calibration/PPSAlCaRecoProducer/ALCARECOPPSCalMaxTracksExpress_cff.py
 existing object defined in: /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-01-23-2300/python/EventFilter/CTPPSRawToDigi/ctppsDiamondRawToDigi_cfi.py
```

Reply:
That error means that the `process` already had a module with label `ctppsDiamondRawToDigi` (that is used in `RawToDigiTask`). If you need the same module here with a different configuration, you need to pick a different label for it.

Reply:
If I switch to completely local objects (with `_` prefix) and follow the same advice, the "override definition" error is gone and a different problem appears.

In that case I use following imports:

```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTriggerRawToDigi as _totemTriggerRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemRPRawToDigi as _totemRPRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsDiamondRawToDigi as _ctppsDiamondRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis

from RecoPPS.Configuration.recoCTPPS_cff import *

_totemTriggerRawToDigi = _totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
_totemRPRawToDigi = _totemRPRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
_ctppsDiamondRawToDigi = _ctppsDiamondRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
_totemTimingRawToDigi = _totemTimingRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
_ctppsPixelDigis = _ctppsPixelDigis.clone(inputLabel = cms.InputTag('hltPPSCalibrationRaw'))

_ctppsRawToDigiTask = cms.Task(
  _totemTriggerRawToDigi,
  _totemRPRawToDigi,
  _ctppsDiamondRawToDigi,
  _totemTimingRawToDigi,
  _ctppsPixelDigis
)
```

The reported problem is:
```python
Step: ALCAPRODUCER Spec: ['PPSCalTrackBasedSel']
(...)
RuntimeError: module has no label.  Perhaps it wasn't inserted into the process?
(...)
RuntimeError: An entry in sequence pathALCARECOPPSCalMaxTracksExpress has no label
```

This and the comment above is related to the following matrix tests:
```bash
runTheMatrix.py -l 1041 --ibeos
```

Reply:
Finally adding `setLabel` solves the problem.
BTW - do you know if there is somewhere good documentation of best practices in dealing with imports and structure of Python configs? Going through the source code of https://github.com/cms-sw/cmssw/tree/master/FWCore/ParameterSet/python doesn't seem to be the most efficient way.

Reply:
> That error means that the `process` already had a module with label `ctppsDiamondRawToDigi` (that is used in `RawToDigiTask`). 

That is correct, most probably in some other piece of configuration a line like: `from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *` appeared.
I suspect this file: https://github.com/cms-sw/cmssw/blob/CMSSW_12_3_X/CalibPPS/TimingCalibration/python/ALCARECOPPSCalTrackBasedSel_cff.py
which was recently added to `autoAlca.py` in https://github.com/cms-sw/cmssw/pull/36702 and included in this PR after squash + rebase.

I believe such import of everything looks like a bad practice in that situation, therefore I asked what are the recommendations in https://github.com/cms-sw/cmssw/pull/36273#discussion_r792059710

> If you need the same module here with a different configuration, you need to pick a different label for it.

What is the suggested way of changing the label?
If I simply use a different variable it doesn't help, see https://github.com/cms-sw/cmssw/pull/36273#discussion_r792054608

Reply:
The python configuration is documented here https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideAboutPythonConfigFile

The problem with this approach is that the "local objects" (should) never end up attached in the `process` in `process.load(...)` call. Things may somehow seem to work (as in "there are no clear errors") because of using same labels as some other modules in the `process`. I'd guess in this case the original modules get run, and given the cloning and customization of parameters that is likely not what was intended.

Reply:
> > If you need the same module here with a different configuration, you need to pick a different label for it.
> 
> What is the suggested way of changing the label? If I simply use a different variable it doesn't help, see [#36273 (comment)](https://github.com/cms-sw/cmssw/pull/36273#discussion_r792054608)

Using a different variable name is exactly the way to assign a different label for a module. If I understood [#36273 (comment)](https://github.com/cms-sw/cmssw/pull/36273#discussion_r792054608) correctly, i.e.
```py
_totemTriggerRawToDigi = _totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
```
the error message suggests `module has no label.  Perhaps it wasn't inserted into the process?` because `_totemTriggerRawToDigi` was not inserted into the `process` because of the leading underscore.

Technically the label needs to be unique. I'd suggest to add some postfix (or prefix) to them, e.g. for testing
```py
totemTriggerRawToDigiTmp = _totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemRPRawToDigiTmp = _totemRPRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsDiamondRawToDigiTmp = _ctppsDiamondRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemTimingRawToDigiTmp = _totemTimingRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsPixelDigisTmp = _ctppsPixelDigis.clone(inputLabel = cms.InputTag('hltPPSCalibrationRaw'))
```
But please use something more descriptive than the `Tmp` above in the end (perhaps @cms-sw/alca-l2 could suggest, e.g. if there is an existing convention in AlCaReco files).

Reply:
Thank you for the last comment, that explains a lot, especially on the leading underscore. 
I wrongly understood, that `_` matters only for the objects imported from the config like `seqALCARECOPPSCalMaxTracksRecoPrompt` and not for all other variables, members of Tasks and Sequences.

I will wait for AlCa suggestion, if there is none, I'd use something prefix like `AlCaRecoProducer` to clarify where the objects belong.

Reply:
@makortel  I've tried the approach with suffix, as you suggested and this ended up in another problem:

```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTriggerRawToDigi as _totemTriggerRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemRPRawToDigi as _totemRPRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsDiamondRawToDigi as _ctppsDiamondRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis

from RecoPPS.Configuration.recoCTPPS_cff import *

totemTriggerRawToDigiAlCaRecoProducer = _totemTriggerRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemRPRawToDigiAlCaRecoProducer = _totemRPRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsDiamondRawToDigiAlCaRecoProducer  = _ctppsDiamondRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
totemTimingRawToDigiAlCaRecoProducer = _totemTimingRawToDigi.clone(rawDataTag = cms.InputTag('hltPPSCalibrationRaw'))
ctppsPixelDigisAlCaRecoProducer = _ctppsPixelDigis.clone(inputLabel = cms.InputTag('hltPPSCalibrationRaw'))

ctppsRawToDigiTaskAlCaRecoProducer = cms.Task(
  totemTriggerRawToDigiAlCaRecoProducer,
  totemRPRawToDigiAlCaRecoProducer,
  ctppsDiamondRawToDigiAlCaRecoProducer,
  totemTimingRawToDigiAlCaRecoProducer,
  ctppsPixelDigisAlCaRecoProducer
)

ctppsRawToDigiAlCaRecoProducer = cms.Sequence(ctppsRawToDigiTaskAlCaRecoProducer)

seqALCARECOPPSCalMaxTracksRecoPrompt = cms.Sequence( ALCARECOPPSCalMaxTracksFilterPrompt + ctppsRawToDigiAlCaRecoProducer  + recoCTPPS )
```

The error message after running: `runTheMatrix.py -l 1041 --ibeos` is:
```python
Step: ALCAPRODUCER Spec: ['PPSCalTrackBasedSel']
(...)
RuntimeError: module has no label.  Perhaps it wasn't inserted into the process?
(...)
RuntimeError: An entry in sequence pathALCARECOPPSCalMaxTracksPrompt has no label
```

Reply:
Hmm, that was unexpected. I don't see any obvious problem by eye. Unfortunately this is a rather annoying problem to debug. A straightforward way would be to comment out elements on Sequences and Tasks (e.g. one by one, or similar to binary search) to find out which one(s) cause the error.

Reply:
It seems that with careful refactoring I was able to adjust the importing schema, see https://github.com/cms-sw/cmssw/pull/36273/commits/01f59d91a1dbf8c20e3cbfee5d3669a8cda50291

Does it comply with standards now ?

Reply:
Thanks! It looks technically correct now, but I'm wondering the intent of the fragment. For example, it defines a producer with a label `totemTriggerRawToDigiAlCaRecoProducerExpress`, but I don't see anything consuming data products with that label. I'd guess the intention would be something different.